### PR TITLE
Clean constants

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

remove
```
// DeprecatedGatewayNameLabel indicates the gateway managing a particular proxy instances. Only populated for Gateway API gateways
DeprecatedGatewayNameLabel = "istio.io/gateway-name"
```

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/config/constants/constants.go [hunk 1]
```diff
@@ -175,8 +172,6 @@ const (
 	RemoteGatewayClassName   = "istio-remote"
 	WaypointGatewayClassName = "istio-waypoint"
 
-	// DeprecatedGatewayNameLabel indicates the gateway managing a particular proxy instances. Only populated for Gateway API gateways
-	DeprecatedGatewayNameLabel = "istio.io/gateway-name"
```
Comment (hanxiaop) [inline pkg/config/constants/constants.go:179]: How about the labels of k8s gateways in `manifests/charts/istio-control/istio-discovery/files`?

File: pilot/pkg/bootstrap/config_compare.go [hunk 2]
```diff
@@ -27,11 +27,8 @@ import (
 // to be triggered. This is to avoid unnecessary pushes only when labels have changed
 // for example.
 func needsPush(prev config.Config, curr config.Config) bool {
-	if prev.GroupVersionKind != curr.GroupVersionKind {
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/config_compare.go:30]: Why remove? May be fine if it 'never happens', but might be good to keep for safety. Or at least seems unrelated to cleaning constants

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 3]
```diff
@@ -618,16 +618,6 @@ func (d *DeploymentController) setGatewayNameLabel(ti *TemplateInput) {
...
-
-	// Base label choice on the deployment's selector
-	_, exists := dep.(*appsv1.Deployment).Spec.Selector.MatchLabels[constants.DeprecatedGatewayNameLabel]
-	if !exists {
-		// The old label doesn't already exist on the deployment; use the new label
-		return
-	}
-
-	// The old label exists on the deployment; use the old label
-	ti.GatewayNameLabel = constants.DeprecatedGatewayNameLabel
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:630]: This is for backwards compatibility from a pretty recent change, not sure we are ready to remove it? @keithmattix

File: pilot/pkg/config/kube/ingress/controller.go [hunk 4]
```diff
@@ -196,15 +196,11 @@ func (c *controller) onEvent(item types.NamespacedName) error {
 		Name:             item.Name + "-" + "virtualservice",
 		Namespace:        item.Namespace,
 		GroupVersionKind: gvk.VirtualService,
-		// Set this label so that we do not compare configs and just push.
-		Labels: map[string]string{constants.AlwaysPushLabel: "true"},
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingress/controller.go:200]: Why is this not needed? It seems it was added for a reason and documented why; this seems to break that?

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pilot/pkg/bootstrap/config_compare.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/config_compare.go:30]: Can not happen, so eliminate the compare

File: pilot/pkg/config/kube/ingress/controller.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:200]: If you look at the `needsPush`, for any non istio config, we return true.

Comment (howardjohn) [reply pilot/pkg/config/kube/ingress/controller.go:200]: this is a virtual service, isn't that istiod config?

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:200]: Oh, my mistake. Just you can search this label, not read at any place except setting it.

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:200]: Originally i did remember it was useful. not sure from when it was not

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:200]: Do you still remember why do we need to always push when ingress update?

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:200]: @howardjohn By reading the ctx, i know the reason. Will fix ingress handler

File: pkg/config/constants/constants.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/config/constants/constants.go:179]: I think we can remove too, as `gateway.networking.k8s.io/gateway-name` is also added

File: pilot/pkg/config/kube/ingress/controller.go [hunk 5]
```diff
@@ -110,10 +113,12 @@ func NewController(client kube.Client, meshWatcher mesh.Holder,
 		classes:      classes,
 		services:     services,
 	}
-	c.queue = controllers.NewQueue("ingress",
-		controllers.WithReconciler(c.onEvent),
-		controllers.WithMaxAttempts(5))
-	c.ingress.AddEventHandler(controllers.ObjectHandler(c.queue.AddObject))
+	c.queue = queue.NewQueueWithID(time.Second, "ingress")
+	c.ingress.AddEventHandler(controllers.FromEventHandler(func(o controllers.Event) {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingress/controller.go:117]: This seems like a regression. We have been intentionally moving controllers _to_ the `controllers.Queue` from the `queue.Instance`, to more closely align with Kubernetes best practice on controller reconciliation.

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:117]: Thankfully, ingress controller stores old ingress, so i think can restore to `controllers.Queue`, and rely on the comparition of the new/old ingress

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 3 above)
Comment (keithmattix) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:630]: Yeah this is still a release or 2 too early; it just got merged in 1.20 IIRC

File: pilot/pkg/config/kube/ingress/controller.go [hunk 6]
```diff
@@ -186,29 +187,31 @@ func (c *controller) onEvent(item types.NamespacedName) error {
 	// we should check need process only when event is not delete,
 	// if it is delete event, and previously processed, we need to process too.
 	if event != model.EventDelete {
+		oldIngress := c.ingresses[config.NamespacedName(ing)]
 		shouldProcess := c.shouldProcessIngressUpdate(ing)
 		if !shouldProcess {
 			return nil
 		}
+		// only trigger when real changes were found
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingress/controller.go:195]: This doesn't seem quite right because it doesn't take into account other dependencies? For example, when Service changes, we need to do a push. Even if Ingress doesn't change

Comment (howardjohn) [reply pilot/pkg/config/kube/ingress/controller.go:195]: I also don't get why we can remove the AlwaysPush. The VS and GW we are pushing below have no spec, so won't they just always be skipped?

**Review by howardjohn — CHANGES_REQUESTED** [review]
This seems to break a lot of ingress stuff, where we don't have a lot of test coverage.

File: pilot/pkg/config/kube/ingress/controller.go (hunk 6 above)
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:195]: as to `AlwaysPush`, we have annotation with `istio.io`

As to `This doesn't seem quite right because it doesn't take into account other dependencies?`

Thanks for remind, need to revert it. Seems we have no way to distinguish unless updating the input. Let's keep the original

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go]:
```
	for annotation := range curr.Meta.Annotations {
		if strings.Contains(annotation, "istio.io") {
			return true
		}
	}
``

File: pilot/pkg/config/kube/ingress/controller.go [hunk 7]
```diff
@@ -196,15 +196,13 @@ func (c *controller) onEvent(item types.NamespacedName) error {
...
 		GroupVersionKind: gvk.VirtualService,
-		// Set this label so that we do not compare configs and just push.
-		Labels: map[string]string{constants.AlwaysPushLabel: "true"},
+		Annotations:      map[string]string{constants.InternalRouteSemantics: constants.RouteSemanticsIngress},
 	}
 	gatewaymetadata := config.Meta{
 		Name:             item.Name + "-" + "gateway",
 		Namespace:        item.Namespace,
 		GroupVersionKind: gvk.Gateway,
-		// Set this label so that we do not compare configs and just push.
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingress/controller.go:206]: How is this fixed? I don't get it

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 3 above)
Comment (keithmattix) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:630]: I think I was probably too aggressive; 1 more release should be ok with an upgrade note

File: pilot/pkg/config/kube/ingress/controller.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:206]: @howardjohn Take a look at the below logic, not only ingress but also gateway api go through this comparison
```
func needsPush(prev config.Config, curr config.Config) bool {

    ...
	// If the config is not Istio, let us just push.
	if !strings.HasSuffix(prev.GroupVersionKind.Group, "istio.io") {
		return true
	}
	// If current/previous metadata has "*istio.io" label/annotation, just push
	for label := range curr.Meta.Labels {
		if strings.Contains(label, "istio.io") {   
			return true
		}
	}
	for annotation := range curr.Meta.Annotations {
		if strings.Contains(annotation, "istio.io") {    // Here if we checked an annotation, we will always push
			return true
		}
	}
	for label := range prev.Meta.Labels {
		if strings.Contains(label, "istio.io") {
			return true
		}
	}
	for annotation := range prev.Meta.Annotations {
		if strings.Contains(annotation, "istio.io") {
			return true
		}
	}
	....
	return true
}

```

Comment (howardjohn) [reply pilot/pkg/config/kube/ingress/controller.go:206]: wow that's extremely subtle and likely to break in the future. I really prefer the explicit AlwaysPush...

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:206]: Hmm, you know this stull of inner constant looks very hardcoding. I would like to remove all those by updating `configHandler` with an explicit arg, so we donot need to compare by `needPush` for some resources converted from ingress/gateway-api

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/ingress/controller.go:206]: But currently we cannot gert rid of `InternalRouteSemantics` annotation easily, cause it is used across the code

File: pilot/pkg/bootstrap/config_compare.go [hunk 8]
```diff
@@ -35,27 +36,11 @@ func needsPush(prev config.Config, curr config.Config) bool {
 	if !strings.HasSuffix(prev.GroupVersionKind.Group, "istio.io") {
 		return true
 	}
-	// If current/previous metadata has "*istio.io" label/annotation, just push
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/config_compare.go:38]: This was not only for "AlwaysPush". There are annotations that impact XDS generation. Below we only compare spec, so if a config changes just the annotation nothing happens.

Things like istio.io/dry-run, etc

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pilot/pkg/bootstrap/config_compare.go (hunk 8 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/config_compare.go:38]: why should make dryrun push?

Comment (howardjohn) [reply pilot/pkg/bootstrap/config_compare.go:38]: Any change to inputs that produces a change in outputs should result in a push so things are recomputed. Otherwise the update would be ignored and generate invalid information.

Its not special about "dry-run", just anything where annotation change leads to XDS changes. I recall dry-run was one motivation use

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/config_compare.go:38]: Can you be more explicit, i donot think of a case we should annotate a config to make istiod push.

Originally this function is to mitigate pushed when labels change to a DR. Simutabeously introduce alwaysPush label, but not sure i know why "*istio.io" should be checked if we checked alwaysPush labels explicitly

Comment (howardjohn) [reply pilot/pkg/bootstrap/config_compare.go:38]: We check istio.io only because we know other random annotations are not used in xds generation. It's a hack to improve performance and ignore irrelevant annotations

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/config_compare.go:38]: ..... any example?

Comment (howardjohn) [reply pilot/pkg/bootstrap/config_compare.go:38]: https://github.com/istio/istio/blob/c5463a0b9e381842049a2f2081f361affc2dd0a2/pilot/pkg/security/authz/builder/builder.go#L162

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/config_compare.go:38]: I just learned another annotation, hha.  This makes me hard to do any simplify. 

BTW, if we need to keep the feature or this is a useful feature, i think so.  we should move it under spec.

**Comment by hzxuzhonghu** [comment]
@howardjohn another look please

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:630]: agree with Keith (at the time of the comment being merged, I had worries about 1.22 being too early). Note Keith added the comment. 

This code costs us nothing but changing now will break users

**Comment by costinm** [comment]
On Mon, Apr 15, 2024 at 7:54 PM hzxuzhonghu ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/bootstrap/config_compare.go
> <https://github.com/istio/istio/pull/50170#discussion_r1566644106>:
>
> > @@ -35,27 +36,11 @@ func needsPush(prev config.Config, curr config.Config) bool {
>  	if !strings.HasSuffix(prev.GroupVersionKind.Group, "istio.io") {
>  		return true
>  	}
> -	// If current/previous metadata has "*istio.io" label/annotation, just push
>
> I just learned another annotation, hha. This makes me hard to do any
> simplify.
>
> BTW, if we need to keep the feature or this is a useful feature, i think
> so. we should move it under spec.
>
We should move EVERYTHING under a spec - as either 'supported for users',
'internal' or 'deprecated',
and add some validation that only 'supported' values can be set by users.

Together with the env variables it creates extreme danger for anyone who
attempts to use OPA or other
validations to restrict users from unsupported features.

**Comment by hzxuzhonghu** [comment]
Yes, agree. The path to moving a env/label/annotation to API seems super harder in istio, we can see the number of those stuff is becoming bigger and bigger.

**Comment by hzxuzhonghu** [comment]
I update the description, only remove `DeprecatedGatewayNameLabel = "istio.io/gateway-name"`

**Comment by howardjohn** [comment]
That's the label I *don't* want to remove. The motivation for the change seems to be a TODO comment that both the author and approved agree was a mistake.
